File: CPP/Medium/GroupAnagram.cpp
#include<bits/stdc++.h>
using namespace std;

vector<vector<string>> groupAnagram(vector<string>strs){
    vector<vector<string>> ans;
    map<string,vector<string>> map;
    for(int i=0; i<strs.size(); i++){
        string temp = strs[i];
        sort(temp.begin(), temp.end());
        map[temp].push_back(strs[i]);
    }
    for(auto x:map){
        ans.push_back(x.second);
    }
    return ans;
}

int main(){
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};
    vector<vector<string>>s = groupAnagram(strs);
    return 0;
}

File: JAVA/Medium/GroupAnagram.java
import java.util.*;

class GroupAnagram{

    public static List<List<String>> groupAnagram(String [] strs){
        List<List<String>> ans = new ArrayList<>();
        HashMap<String,List<String>> map = new HashMap<>();
        for(int i=0; i<strs.length; i++){
            char [] temp = strs[i].toCharArray();
            Arrays.sort(temp);
            String s = String.valueOf(temp);
            if(map.get(s)!=null){
                List<String> a = map.get(s);
                a.add(strs[i]);
                map.put(s,a);
            }
            else{
                List<String> a = new ArrayList<>();
                a.add(strs[i]);
                map.put(s,a);
            }
        }
        for(Map.Entry<String, List<String>> x : map.entrySet()){
            ans.add(x.getValue());
        }

        return ans;
    }
    public static void main(String args[]){
        String [] words = {"eat","tea","tan","ate","nat","bat"};
        List<List<String>> res = groupAnagram(words);
        System.out.println(res);
    }
}